### ISSUE TITLE
Should recipes drive the renderer?

### ISSUE THREAD

**Issue by wbamberg** [mdn/stumptown-content#68]
(meta: taken from https://github.com/mdn/stumptown-content/wiki/Notes-from-Whistler-All-Hands,-June-2019#decide-whether-recipes-should-drive-the-renderer)

In stumptown we've expressed a desire to distinguish between (1) "MDN content" which is an abstract collection of web documentation, and (2) "MDN the website" (developer.mozilla.org). The idea is that "MDN the website" is the main consumer of "MDN content", but not necessarily the only one: editors and devtools could also want to integrate the content. This is reflected in the division between "stumptown-content" and "stumptown-renderer".

Maintaining this distinction is tricky, and one of the skirmishes is the meaning of "recipes". These describe what content should be included in a particular type of document (e.g. https://github.com/mdn/stumptown-content/blob/master/recipes/html-element.yaml). These files have - or at least used to have - a sort of double life. On the one hand they describe which bits of content authors should supply when they're documenting a thing. On the other hand they might be a description of how MDN should render pages.

The concrete difference would be seen if, for example, we decided to stop displaying CSS formal syntax in pages, but wanted to keep the content (maybe because other consumers were using it).

We discussed this recently (https://github.com/mdn/stumptown-content/issues/48) and the resolution was that recipes should *not* drive the renderer, but instead we should just hardcode the different content types are React components (https://github.com/mdn/stumptown-renderer/pull/18, which basically implements https://github.com/mdn/stumptown-content/issues/48#issuecomment-498992233). This means that the renderer has nothing to do with recipes.

In Whistler, @peterbe wanted us to reconsider this decision, and make recipes drive the renderer.

Pros: things get simpler. Not just in the case of recipes: there are other places too where this distinction can get messy: an obvious one is sidebars, which seem a very MDN-y thing, but whose content is clearly an authorial choice (so it is very tempting to put them in stumptown-content). It's easier to make design choices to satisfy real consumers than to satisfy theoretical consumers.

Cons: we acknowledge that stumptown-content is being built to serve MDN-the-website. It's still possible for other consumers to use it, and it might still be as easy, but it's certain that we will introduce extra stuff they don't care about (like sidebars) and possible we will make choices about what to expose or how to expose it that are very suboptimal for other consumers.

Either way, we should be very clear about the choice we're making here and its implications.

**Comment by peterbe** [comment]
Coupla thoughts...

* I think the original argument (for reintroducing recipes) is that it's a great chance to A) use it to lint that everything in the packaged JSON is there so that the renderer will work, and B) make life easier for the renderer. 
* Stumptown-content already has some strong opinions about the rendering that it technically could omit and that's the URI (aka. slug aka. URL). One could have argued that stumptown-content doesn't care how that's consumed and it's up to the renderer. But it is very relevant for content editors to influence with this key. 
* We also discussed, a new opportunity in the recipe; if you want to dictate that a section must exist (for linting) but hint that it shouldn't be rendered you could type something like `formal-syntax^`

Perhaps this is too soon for me to bring up but I never liked the connection between content and recipes. It's kinda clunky to go from content to recipes with `../../../../recipes/html/reference/element.yaml`. 
An idea would be to put that stuff in `/content/html/reference/element/recipe.yaml`. Lookup, in stumptown-content scripting, would be something like this:
```javascript
function getRecipe(currentContentPath) {
    let dir = path.dirname(currentContentPath);
    if (path.join(dir, 'recipe.yaml')) {
        return fs.readFileSync(path.join(dir, 'recipe.yaml'))
    }
    if (dir === '/') {
       throw "Can't go any deeper"
    } 
   let parent = path.join(dir, '..')
   return getRecipe(parent);
}
```
(pseudo code but I hope it illustrates)
With this you could have a `recipe.yaml` sprinkled all over the content. You'd be able to have generic ones for a whole batch of documents (e.g. `content/css/reference/`) but also for very specific things.

**Comment by wbamberg** [comment]
I think I'm in favour of having recipes drive the renderer, mostly because I think it'll make things simpler and generally easier to understand.

> An idea would be to put that stuff in /content/html/reference/element/recipe.yaml

I'm not convinced by this for a few reasons.

1. it ties the directory structure under /content to the types of pages, which is much less flexible than the current arrangement. For example, guide pages are a very generic page type, that can occur across lots of different domains. If we have a recipe for them, do we have to keep all guide pages under their own hierarchy? Similarly, we will likely have different page types for Web API methods, properties, and events. So does this mean we must have all events, across all Web APIs, under a single hierarchy, rather than splitting them up by API? We *might* decide anyway that this is a desirable organization (although I think it probably isn't), but it would be nice to make that choice on its own merits rather than being forced into it.

2. it's less explicit. Right now if I'm looking at https://github.com/mdn/stumptown-content/blob/master/content/html/elements/article/meta.yaml I know what kind of page it is, because it tells me. With this proposal I have to look up the directory hierarchy to work it out.

3. recipes are pretty central to stumptown (more so if they drive the renderer) and I really like the idea of having a single place where it's easy to see what all the page types are. In this proposal recipes are as you say sprinkled through the content: it becomes quite hard for a reader to understand "what are all the different page types we support.

On the other side I'm not sure what the benefit of this proposal is. You say it's "kinda clunky" but I'm not really sure what that means in practical terms or what problems it causes.

**Comment by peterbe** [comment]
I forgot to add an argument to my comments above: Having the recipes sit alongside the .md files means less typing. If you open any random `/content/**/*.md` file instead of reading its front-matter or its meta.yaml the recipes always that recipe.yaml of its directory. Or its parent directory. 

**But** as you say, this is less explicit. It'd also make it impossible to have "unnatural" recipes. E.g. if you have `Something/refpage1.md`, `Something/refpage2.md`, `Something/refpage3.md` they could all share `Something/recipe.yaml` implicitly. But if you also have `Something/about-these-refs.md` it probably needs a totally different recipes since, as the name implies, is not a reference page but a general verbiage page. 

**In conclusion; I'm won over by your arguments.** Let's stick to `/recipes/**/*.yaml` and have the `.md` files (or their `meta.yaml`) explicitly point to the recipe file.

**Comment by peterbe** [comment]
> You say it's "kinda clunky" but I'm not really sure what that means in practical terms or what problems it causes.

I think I meant jumping around in the directory structure. It also means that you have to manually type in the name of the recipe file path which could be "automated" by implictness. 

But see my comment above. 

Generally I'm not a huge fan of project structures that are like this:

* Project Root
  * Components
    * Home
    * Dashboard
  * Tests
    * Home
    * Dashboard

I think this is better:

* Project Root
  * Home
    * Component
    * Tests
  * Dashboard
    * Component
    * Test

**Comment by ddbeck** [comment]
If information in the recipes is going to drive the renderer, then I'd like to see the recipe itself get baked into the JSON output from stumptown. I'm actively opposed to using the recipe YAML themselves as a way of communicating anything to any consumer of stumptown. In other words, if recipes are meant to be consumed outside of this repo, then running `build-json` (or, someday, installing the stumptown npm package) should get you all the data meant to be consumed.

I have a few motivations here:

1. Asking consumers to parse YAML pushes some avoidable complexity onto consumers. Lots of languages include a JSON parser in the standard library; few include YAML. YAML is riskier to work with than JSON (e.g., PyYAML and js-yaml are *unsafe by default*).
2. Recipe information can be usefully baked into the resulting JSON without requiring a recipe lookup. For example, the order of content units can be represented as an array. Or the type of content—which recipe it followed—can be part of the resulting JSON.
3. I'd like you to not [make a liar out of me](https://github.com/mdn/stumptown-content/issues/17#issuecomment-495230964).

I'll grant that the features of various recipe types might be nice to have, independently of any one entry (e.g., if you want to know what the required and optional pieces of all the content made with `html-element` and not just `html.elements.div`). But if we intend for it to be used outside of stumptown-content, then let's make that an explicit part of the JSON output.

**Comment by peterbe** [comment]
I think @wbamberg has a patch brewing somewhere that'll make the packaged .json files look something like this:
```json
{
  "data":  {
    "title": "Video ELement",
    ...
  },
  "metadata": {
    "contributors": [...],
    "date": "2019-07-01T12:13:14.123Z",
    "recipe": {
      ...
    }
  }
```

Yes, the `stumptown-renderer` shouldn't need to parse or figure out YAML. It did once in an early prototype because I was too eager to get to it before it got packaged into the .json files. 
Ideally, the whole `packaged` directory is all stumptown-renderer should want/need.

**Comment by wbamberg** [comment]
So I think when we say "the recipe should drive the renderer" we mean something like this.

At the moment, the renderer includes a bunch of "low-level" atomic components, that render bits of a page (like [a thing that renders an interactive example](https://github.com/mdn/stumptown-renderer/blob/master/client/src/ingredients/interactive-example.js)). These atomic components are likely to be used by many different page types.

But then for each page type we need a way to describe which "bits of a page" get rendered, and in what order: that is, how these atomic components get marshalled to render a complete page. At the moment this is hardcoded for each page type, like this: https://github.com/mdn/stumptown-renderer/blob/master/client/src/document.js#L139-L168.

When we say, "the recipe should drive the renderer", I think we mean that instead of having this hardcoded thing, then the renderer reads the recipe for the page type, and used this to figure out which atomic components to call and in what order. So it looks at, say, https://github.com/mdn/stumptown-content/blob/master/recipes/html-element.yaml (or rather, the JSONified version of that) and walks through it calling the appropriate atomic component for each item.

So the order of items in the recipe becomes significant.

>  if you want to dictate that a section must exist (for linting) but hint that it shouldn't be rendered you could type something like formal-syntax^

This feels like too much of a close link between the recipe and the renderer. An alternative could be: the "atomic component" for an item is just a no-op for formal syntax, if we decide not to show it in the page. Or, we don't have an atomic component for it, so the renderer's behavior is just: "if I don't have an atomic component for an item in the recipe, skip it".

But that's just my opinion and isn't founded on anything very concrete.

> Recipe information can be usefully baked into the resulting JSON without requiring a recipe lookup. For example, the order of content units can be represented as an array.

This seems like it will feel odd for other consumers. Other consumers can just ignore recipes, but if the JSON object's organization just is the page organization, then they can't escape MDN's rendering choices. For example, at the moment we have an object `prose` that contains all the prose sections:

```
{
  "title": "<abbr>: The Abbreviation element",
  "prose": {
     "short_description": "The abbr element is very interesting",
     "overview": "more stuff about it"
  }
}
```

If the JSON object's structure must be the page structure, we have to do something like this:

```
{
  "title": "<abbr>: The Abbreviation element",
  "prose_short_description: "The abbr element is very interesting",
  "interactive_example": ...
  "prose_overview": "more stuff about it",
  ...
}
```

> Or the type of content—which recipe it followed—can be part of the resulting JSON.

That's more what I was thinking. I don't know if it should point to a separate recipe file or embed the recipe.

I'd love to hear more thoughts from @Elchi3 and @ddbeck on this though.

**Comment by ddbeck** [comment]
> If the JSON object's structure must be the page structure, we have to do something like \<snip\>

That's not really what I was getting at. For starters, that structure doesn't show the order of the recipe (keep in mind that JSON objects are unordered). I was thinking you would do something like this, where the JSON contains an array of the atomic components:

```
{
  "title": "<abbr>: The Abbreviation element",
  "type": "html_element",
  "content": [
    {
      "type": "prose",
      "id": "short_description",
      "text": "The abbr element is very interesting"
    },
    {
      "type": "interactive_example",
      …
    },
    {
      "type": "prose",
      "id": "overview",
      "text": "more stuff about it"
    }
    {
      "type": "prose",
      "id": "attributes_text",
      "text": null  /* an optional prose section is empty in this example; there are other ways to do this */
    }
    …
  ],
  "related_content": …
}
```

For MDN, the renderer steps through the content and decides whether it can or should render that object. For other consumers, it's a little more cumbersome to get some content (e.g., if you want just the short description, you'll have to filter the array for that object), but I don't think it's particularly hard to reason about.

**Comment by peterbe** [comment]
If what @ddbeck described is feasible, it certainly seems very straight-forward to render that. 

The only thing that stands out, and I feel bad for "nit picking" an example, but what's the point of the `"type": "html_element"`?

**Comment by ddbeck** [comment]
> The only thing that stands out, and I feel bad for "nit picking" an example, but what's the point of the `"type": "html_element"`?

Oh that's just the recipe (if we wanted to tell consumers what the recipe is, that is). I should come up with better names for things than "type" because I'm inclined to use it everywhere.

**Comment by wbamberg** [comment]
Thanks for clarifying, Daniel. I was just being inattentive not making the JSON an array, but I like your more fleshed-out example better anyway.

One thing I like about it is that it seems to fit better with what we would do for guide pages, which don't have a strict structure (in that each guide page might have a different set of component pieces, in a different order), so they will anyway need to be represented somewhat like:

```
- some prose
- a live example
- some more prose
- a live example
- yet more prose
- a compat table
- more prose again
```

...and again, the renderer walks through this feeding each piece into a rendering component.

**Comment by wbamberg** [comment]
I think [Daniel's suggestion](https://github.com/mdn/stumptown-content/issues/68#issuecomment-516347924) makes sense but this is a big change with a major bearing on the external interface of stumptown-content. It's also a decision we've made (one way) and are now revising.

It feels like this design ties stumptown much closer to the rendered pages on MDN. I don't think that's necessarily a bad thing but it is a thing :).

But I don't think it matter much *at this early stage* which we choose. We can change our minds once we have more experience of working with this. So I think the main thing I want to avoid now is spending a lot of time on issues that don't really advance the experiment (compared to, say, being able to render guide pages or sidebars or landing pages), and on which we don't yet have enough real-world experience to make sensible judgements about.

So in short, I guess I think: "at this stage, we can do either, but let's not spend all our time second-guessing our choices".

**Comment by ddbeck** [comment]
Yep, it seems like not letting the renderer have knowledge of information in the recipe is adding friction to our overall experiment, so let's try on this ordered representation instead.

**Comment by wbamberg** [comment]
Closed via https://github.com/mdn/stumptown-content/pull/107.
